x-pack/plugins/rule_registry/server/utils/create_get_summarized_alerts_fn.ts

```diff
@@ -0,0 +1,359 @@
+    }
+    const queryByExecutionUuid: boolean = !!executionUuid;
+    const queryByTimeRange: boolean = !!start && !!end;
+    // Either executionUuid or start/end dates must be specified, but not both
+    if (
+      (!queryByExecutionUuid && !queryByTimeRange) ||
+      (queryByExecutionUuid && queryByTimeRange)
+    ) {
+      throw new Error(
+        `Must specify either execution UUID or time range for summarized alert query.`
+      );
+    }
+
+    // Get the rule data client reader
+    const { ruleDataClient, useNamespace } = opts;
+    const ruleDataClientReader = useNamespace
+      ? ruleDataClient.getReader({ namespace: spaceId })
+      : ruleDataClient.getReader();
+
+    if (queryByExecutionUuid) {
+      return await getAlertsByExecutionUuid({
+        ruleDataClientReader,
+        ruleId,
+        executionUuid: executionUuid!,
+        isLifecycleAlert: opts.isLifecycleAlert,
+      });
+    }
+
+    return await getAlertsByTimeRange({
+      ruleDataClientReader,
+      ruleId,
+      start: start!,
+      end: end!,
+      isLifecycleAlert: opts.isLifecycleAlert,
+    });
+  };
+
+interface GetAlertsByExecutionUuidOpts {
+  executionUuid: string;
+  ruleId: string;
+  ruleDataClientReader: IRuleDataReader;
+  isLifecycleAlert: boolean;
+}
+
+const getAlertsByExecutionUuid = async ({
+  executionUuid,
+  ruleId,
+  ruleDataClientReader,
+  isLifecycleAlert,
+}: GetAlertsByExecutionUuidOpts) => {
+  if (isLifecycleAlert) {
+    return getLifecycleAlertsByExecutionUuid({ executionUuid, ruleId, ruleDataClientReader });
+  }
+
+  return getPersistentAlertsByExecutionUuid({ executionUuid, ruleId, ruleDataClientReader });
+};
+
+interface GetAlertsByExecutionUuidHelperOpts {
+  executionUuid: string;
+  ruleId: string;
+  ruleDataClientReader: IRuleDataReader;
+}
+
+const getPersistentAlertsByExecutionUuid = async <TSearchRequest extends ESSearchRequest>({
+  executionUuid,
+  ruleId,
+  ruleDataClientReader,
+}: GetAlertsByExecutionUuidHelperOpts) => {
+  // persistent alerts only create new alerts so query by execution UUID to
+  // get all alerts created during an execution
+  const request = getQueryByExecutionUuid(executionUuid, ruleId);
+  const response = (await ruleDataClientReader.search(request)) as ESSearchResponse<
+    AlertDocument,
+    TSearchRequest
+  >;
+
+  return {
+    new: getHitsWithCount(response),
+    ongoing: {
+      count: 0,
+      alerts: [],
+    },
+    recovered: {
+      count: 0,
+      alerts: [],
+    },
+  };
+};
+
+const getLifecycleAlertsByExecutionUuid = async ({
+  executionUuid,
+  ruleId,
+  ruleDataClientReader,
+}: GetAlertsByExecutionUuidHelperOpts) => {
+  // lifecycle alerts assign a different action to an alert depending
+  // on whether it is new/ongoing/recovered. query for each action in order
+  // to get the count of each action type as well as up to the maximum number
+  // of each type of alert.
+  const requests = [
```

Comment:
any chance we could do this with a single agg?

Reply:
I don't believe with a `terms` aggregation, we can get the source documents for the hits that fall into a bucket (please correct me if I'm wrong!). It looks like we might be able to do it with a `top_hits` aggregation but it says in [the docs ](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-top-hits-aggregation.html) that `top_hits` is not recommended as a top level aggregation, which is how we would want to use it.

Reply:
Ah, makes sense, thx!